### PR TITLE
Invalidate process cache on deployment rejection

### LINKED ISSUE #13254: When the deployment of a process definition failed, clients can still start new process instances of the actually non-existing process definition
**Describe the bug**

When the command to deploy processes failed with, for example, `EXCEEDED_BATCH_RECORD_SIZE`, any processes - that have been cached during the deployment process - are still kept in cache afterward. That way, clients can start process instances of these processes. Once a leader change happened (or the broker restarted), the Stream Process fails to replay these started process instances:

```
java.lang.RuntimeException: java.lang.IllegalStateException: Expected to find a process deployed with key '2251799813685405' but not found.
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$replayNextEvent$4(ReplayStateMachine.java:169) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.future.FutureContinuationRunnable.run(FutureContinuationRunnable.java:33) ~[zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:94) ~[zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorJob.execute(ActorJob.java:45) [zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorTask.execute(ActorTask.java:119) [zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorThread.executeCurrentTask(ActorThread.java:106) [zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorThread.doWork(ActorThread.java:87) [zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorThread.run(ActorThread.java:198) [zeebe-scheduler-8.2.4.jar:8.2.4]
Caused by: java.lang.IllegalStateException: Expected to find a process deployed with key '2251799813685405' but not found.
	at io.camunda.zeebe.engine.state.deployment.DbProcessState.getFlowElement(DbProcessState.java:288) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.engine.state.appliers.ProcessInstanceElementActivatingApplier.createEventScope(ProcessInstanceElementActivatingApplier.java:261) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.engine.state.appliers.ProcessInstanceElementActivatingApplier.applyState(ProcessInstanceElementActivatingApplier.java:51) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.engine.state.appliers.ProcessInstanceElementActivatingApplier.applyState(ProcessInstanceElementActivatingApplier.java:32) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.engine.state.appliers.EventAppliers.applyState(EventAppliers.java:302) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.engine.Engine.replay(Engine.java:110) ~[zeebe-workflow-engine-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$replayEvent$7(ReplayStateMachine.java:230) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at java.util.Optional.ifPresent(Unknown Source) ~[?:?]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.replayEvent(ReplayStateMachine.java:230) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at java.util.Iterator.forEachRemaining(Unknown Source) ~[?:?]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$tryToReplayBatch$5(ReplayStateMachine.java:208) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.db.impl.rocksdb.transaction.ZeebeTransaction.run(ZeebeTransaction.java:84) ~[zeebe-db-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.tryToReplayBatch(ReplayStateMachine.java:206) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.stream.impl.ReplayStateMachine.lambda$replayNextEvent$3(ReplayStateMachine.java:165) ~[zeebe-stream-platform-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.retry.ActorRetryMechanism.run(ActorRetryMechanism.java:28) ~[zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.retry.RecoverableRetryStrategy.run(RecoverableRetryStrategy.java:48) ~[zeebe-scheduler-8.2.4.jar:8.2.4]
	at io.camunda.zeebe.scheduler.ActorJob.invoke(ActorJob.java:92) ~[zeebe-scheduler-8.2.4.jar:8.2.4]
	... 5 more
```

Basically, when processing a deployment command the following happens

1. `DeploymentCreateProcessor#processRecord()` transforms the given resources accordingly
https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java#L129-L135
2. If succeeded, the engine will create timers for any timer start event
https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java#L137-L142
3. When executing the `#createTimerIfTimerStartEvent()` it will try to get the process definition by key:
https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java#L193-L203
4. Getting the process definition by key will eventually put the process definition into a map `processesByProcessIdAndVersion` for caching purposes:
https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java#L201-L211
5. Once `#createTimerIfTimerStartEvent()` finishes, the engine will write a follow-up event to indicate the deployment has been created
https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java#L146
6. Writing this follow-up event might fail with `EXCEEDED_BATCH_RECORD_SIZE`

As a consequence, the RocksDB transaction gets rolled back, meaning, the process definitions are not stored in RocksDB. But the not successful deployed process definitions are still present in the map `processesByProcessIdAndVersion`. That way, when a client starts a process by `bpmnProcessId`, the engine will get the actually non-existing process definition from the map `processesByProcessIdAndVersion` and start a process instance eventually. And as long as the leader keeps the same and Zeebe does not restart, process instances of the non-existing process definition can get started.

Additionally, the `ProcessVersionManager#versionCache` caches the latest version of a process definition. The same applies here, when the deployment fails, the cached latest version has been already increased and stays present in the cache. Basically, since `processesByProcessIdAndVersion` contains a non-existing process definition (for example, with version `12`) and the `versionCache` points to the non-existing version `12` of that process definition, the engine will return the process definition from `processesByProcessIdAndVersion`

https://github.com/camunda/zeebe/blob/e4c88b2a20f48e764118caefa5052199e12ad15e/engine/src/main/java/io/camunda/zeebe/engine/state/deployment/DbProcessState.java#L214-L231

Before `versionCache` have been introduced, the engine would get the latest version from RocksDB which would not be `12` for sure. That way, the engine ensured that it always returns a process definition that actually exists.

**To Reproduce**
1. Deploy many processes at once (containing a process with `bpmnProcessId=foo` )
2. Ensure that the deployment fails with `EXCEEDED_BATCH_RECORD_SIZE`
3. Start a process instance by `bpmnProcessId=foo`  => The process instance gets started accordingly.
4. Stop the broker
5. Start the broker

**Expected behavior**
* Whenever a deployment fails, the client cannot start any process instance of any of those process definitions.
* To be more concrete, any process definitions that get cached when processing the deployment command must be removed when the deployment failed.
* Or as an alternative, don't cache any process definitions during deployment.
* The same applies to `ProcessVersionManager#versionCache`

**What is the impact of the issue?**
* Any follower cannot replay any events of process instances belonging to a non-existing process definition. That way, the stream processor will stop replaying anything. The follower won't create any new snapshots, and so on.
* If there is a failover, the new leader will fail to replay the events when recovering from the last snapshot. It will again stop the stream processor. No snapshots are taken anymore. The back pressure goes up to 100%. The partition gets unhealthy.
* It may result in process definitions that are available on partition >= 2 but not on partition 1. For example:
  1. Deploy the first version of a process `foo` that succeeds
  2. Deploy a new version of the process `foo` that fails with `EXCEEDED_BATCH_RECORD_SIZE` (try doing multiple times)
  3. Then redeploy the first version once again (and let it succeed)
  4. => It will result in a process definition with version `x` (`x > 1`) which gets distributed to other partitions but on partition 1 this process definition won't get persisted in RocksDB.
  5. So on partition > 1, there are running process instances referencing process definition with version `x`. But these process instances cannot be opened in Operate, because the process definition was never exported and imported.
* There is no way to recover from that. Only purging the clusters resolves the issue.

**Environment:**
- Zeebe Version: 8.2.4

---

related to #5723 
related to [SUPPORT-17483](https://jira.camunda.com/browse/SUPPORT-17483)
related to #12873

### PR DESCRIPTION
**Pull request by korthout**
## Description

This resolves a critical bug where process instances can be created for processes from rejected deployments. This could lead to unrecoverable partitions.

The problem was that a cache is build up for each of the processes during the processing of the `Deployment:CREATE` command, but this cache was not invalidated when that command is rejected.

This solution only provides a quick fix by completely clearing the cache, which comes at some performance loss when deployments are rejected.

In a future iteration, we'll need to investigate other solutions for the long term. We should also remove the cache entirely from any operations in the immutable ProcessState.

## Related issues

closes #13254

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

**Comment by korthout** [comment]
Sorry about the force-push. I just had to remove an incorrect import in the test case. I've amended the commit and force-pushed it.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java [hunk 1]
```diff
@@ -96,6 +96,9 @@ public void processRecord(final TypedRecord<DeploymentRecord> command) {
     try {
       createTimerIfTimerStartEvent(command);
     } catch (final RuntimeException e) {
+      // Make sure the cache does not contain any leftovers from this run (by hard resetting)
+      processState.clearCache();
+
       final String reason = String.format(COULD_NOT_CREATE_TIMER_MESSAGE, e.getMessage());
```
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentCreateProcessor.java:102]: :thought_balloon: Wondering we need this block at all, or whether we can handle this by our onProcessingError handle mechanism.

Nothing what you need to do now ofc

File: engine/src/main/java/io/camunda/zeebe/engine/state/immutable/ProcessState.java [hunk 2]
```diff
@@ -30,4 +30,7 @@ public interface ProcessState {
 
   <T extends ExecutableFlowElement> T getFlowElement(
       long processDefinitionKey, DirectBuffer elementId, Class<T> elementType);
+
+  /** TODO: Remove the cache entirely from the immutable state */
+  void clearCache();
```
Comment (Zelldon) [inline engine/src/main/java/io/camunda/zeebe/engine/state/immutable/ProcessState.java:35]: :cry:

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java [hunk 3]
```diff
@@ -294,4 +294,37 @@ public void shouldDoAtomicDeployments() {
             tuple(DeploymentIntent.CREATED, RecordType.EVENT),
             tuple(DeploymentDistributionIntent.DISTRIBUTING, RecordType.EVENT));
   }
+
+  /** Regression test against https://github.com/camunda/zeebe/issues/13254 */
+  @Test
```
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java:299]: :wrench: 
```suggestion
  @RegressionTest("https://github.com/camunda/zeebe/issues/13254"
```

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java [hunk 4]
```diff
@@ -294,4 +294,37 @@ public void shouldDoAtomicDeployments() {
...
+  public void shouldNotCacheProcessesWhenDeploymentRejected() {
+    // given
+    final BpmnModelInstance invalidProcess =
+        Bpmn.createExecutableProcess("too_large_process")
+            .startEvent()
+            .documentation("x".repeat((int) ByteValue.ofMegabytes(3)))
```
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java:305]: :wrench: 
```suggestion
            // In order to cause BATCH SIZE EXCEEDING we add a big comment
            .documentation("x".repeat((int) ByteValue.ofMegabytes(3)))
```

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java [hunk 5]
```diff
@@ -294,4 +294,37 @@ public void shouldDoAtomicDeployments() {
             tuple(DeploymentIntent.CREATED, RecordType.EVENT),
             tuple(DeploymentDistributionIntent.DISTRIBUTING, RecordType.EVENT));
   }
+
+  /** Regression test against https://github.com/camunda/zeebe/issues/13254 */
+  @Test
+  public void shouldNotCacheProcessesWhenDeploymentRejected() {
```
Comment (Zelldon) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/DeploymentRejectionTest.java:300]: :wrench: 
```suggestion
  public void shouldNotBeAbleToCreateInstanceWhenDeploymentIsRejected() {
```

**Review by ChrisKujawa — APPROVED** [review]
Thanks @korthout for the fast fix. :rocket:

 I hope we take the time still afterward to clean up this state, and related issues. :)

**Review by romansmirnov — APPROVED** [review]
LGTM! Thanks for acting so quickly on this bug!

**Review by remcowesterhoud — APPROVED** [review]
LGTM! 🚀

**Comment by korthout** [comment]
bors merge

**Comment by remcowesterhoud** [comment]
@korthout let me save you the time of investigating

The problem here is that `@RegressionTest` isn't Junit4 compatible. It adds the the Junit5 `@Test` annotation instead of 4. As a result the rules don't work which is why the `streamProcessingComposite` is null

**Comment by korthout** [comment]
bors merge

**Pull request merged**